### PR TITLE
Retail-like quest Tragedy In Von Hellman Forest (22)

### PR DESCRIPTION
**Pull request by Joxit**

### CODE REVIEW

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+			if (qs.isMemoState(10))
+			{
+				qs.setMemoState(11);
+			}
+			else if (npc.getScriptValue() == 1)
+			{
+				takeItems(attacker, JEWEL_OF_ADVENTURER_1, -1);
+				giveItems(attacker, JEWEL_OF_ADVENTURER_2, 1);
+				qs.setCond(11, true);
+				qs.unset("memoState");
```
Comment (Zoey76) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: Memo state should be updated to use new methods.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+				}
+			}
+		}
+		return super.onKill(npc, killer, isSummon);
+	}
+	
+	@Override
+	public String onTalk(L2Npc npc, L2PcInstance talker)
+	{
+		final QuestState qs = getQuestState(talker, false);
```
Comment (Zoey76) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: It should probably be true.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 3]
```diff
@@ -0,0 +1,659 @@
...
+				final int cond = qs.getCond();
+				if (((5 <= cond) && (cond <= 7)) && hasQuestItems(player, CROSS_OF_EINHASAD) && hasQuestItems(player, LOST_SKULL_OF_ELF))
+				{
+					if (_tifarenOwner == 0)
+					{
+						_tifarenOwner = player.getObjectId();
+						final L2Npc ghost2 = addSpawn(GHOST_OF_PRIEST, PRIEST_LOC, true, 0);
+						ghost2.setScriptValue(player.getObjectId());
+						qs.startQuestTimer("DESPAWN_GHOST2", 1000 * 120, ghost2);
+						ghost2.broadcastPacket(new NpcSay(ghost2.getObjectId(), 0, ghost2.getId(), NpcStringId.DID_YOU_CALL_ME_S1).addStringParameter(player.getName()));
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: Use Say2.NPC_SAY as the 2nd parameter for NpcSay.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+				}
+				break;
+			}
+			case "31528-04.html":
+			{
+				playSound(player, QuestSound.AMBSOUND_HORROR_03);
+				htmltext = event;
+				break;
+			}
+			case "31528-08.html":
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: event is called "x.html", but does not return any html?

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+					else
+					{
+						htmltext = "31334-03.html";
+					}
+				}
+				break;
+			}
+			case "31334-04.html":
+			{
+				qs.startQuest();
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: This also needs a `if (qs.isCreated())` check.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+				htmltext = event;
+				break;
+			}
+			case "31334-06.html":
+			{
+				if (hasQuestItems(player, CROSS_OF_EINHASAD))
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: This needs a cond check wrapper, same for many other cases below this.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 7]
```diff
@@ -0,0 +1,659 @@
...
+	{
+		final QuestState qs = getQuestState(attacker, false);
+		
+		if ((qs != null) && qs.isCond(10) && hasQuestItems(attacker, JEWEL_OF_ADVENTURER_1))
+		{
+			if (qs.isMemoState(10))
+			{
+				qs.setMemoState(11);
+			}
+			else if (npc.getScriptValue() == 1)
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: `isScriptValue(1)`

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 8]
```diff
@@ -0,0 +1,659 @@
...
+		{
+			if (Util.checkIfInRange(1500, killer, npc, true))
+			{
+				_soulWellNpc = null;
+			}
+		}
+		else
+		{
+			final QuestState qs = getQuestState(killer, false);
+			if ((qs != null) && hasQuestItems(killer, CROSS_OF_EINHASAD) && !hasQuestItems(killer, LOST_SKULL_OF_ELF) && (getRandom(100) < 10))
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: You need to check cond/state here aswell. `qs != null` can still be State.CREATED.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 9]
```diff
@@ -0,0 +1,659 @@
...
+		}
+		return super.onKill(npc, killer, isSummon);
+	}
+	
+	@Override
+	public String onTalk(L2Npc npc, L2PcInstance talker)
+	{
+		final QuestState qs = getQuestState(talker, true);
+		String htmltext = getNoQuestMsg(talker);
+		if (qs == null)
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: qs cannot be null if the 2nd parameter to `getQuestState()` is true.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+		{
+			return htmltext;
+		}
+		switch (qs.getState())
```
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: You can easily drop this switch on state and put the CREATED/COMPLETED blocks inside the current State.STARTED TIFAREN block. Just add a `case 0` and put both of them in there.
And add a cond check for `case GHOST_OF_PRIEST`. But at least then you'll get rid of 2 extra nesting levels that make this code way too wide.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java (hunk 3 above)
Comment (Joxit) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: There is no NPC_SAY in Say2, do you mean NPC_ALL?

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java (hunk 10 above)
Comment (Joxit) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: You are right, my code is a bit wide, but I did my switch like this to better see how the quest is conducted. 
There is no need of a check in ``case GHOST_OF_PRIEST``. Everybody can talk to him, even if the talker is not at this step.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java (hunk 3 above)
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: Eh, that, yes.

File: L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java (hunk 10 above)
Comment (jurchiks) [inline L2J_DataPack/dist/game/data/scripts/quests/Q00022_TragedyInVonHellmannForest/Q00022_TragedyInVonHellmannForest.java]: I believe the primary switch should always be on NPC ID, and the NPCs should be ordered in the order they are introduced to the player throughout the quest. There is no quest which has 2 starting NPCs so having a switch or an IF with State.CREATED for one NPC is pointless.

In any case, the logic goes like this: first you click on an NPC, and THEN the script checks what you're allowed to talk about, not the other way around. That's why NPC ID > cond/state.

**Comment by jurchiks** [comment]
http://pastebin.com/1aJ9T2iB
vs
http://pastebin.com/CRfGjLu3

Aside from getting rid of those 2 nesting levels, the code is shortened by 11 lines and TIFAREN htmls are all in one case, not in 3 different places. No cons, only pros.

**Comment by Joxit** [comment]
Nice job, but it will not add an exploit? If a player has not the variable started but has the condition?

**Comment by jurchiks** [comment]
It's not possible. When quest is not started yet, the variable doesn't exist, and when quest is completed or aborted, all variables except state are deleted.

**Comment by Joxit** [comment]
Okay :)

**Comment by jurchiks** [comment]
Welp, all good from my perspective now :)

**Comment by Sdw-** [comment]
Can someone check it please ?

**Comment by Joxit** [comment]
Commits cleaned and tested :)

**Pull request merged**